src/frontend/src/optimizer/plan_node/stream_asof_join.rs

```diff
@@ -0,0 +1,350 @@
+use crate::optimizer::plan_node::utils::IndicesDisplay;
+use crate::optimizer::plan_node::{EqJoinPredicate, EqJoinPredicateDisplay};
+use crate::optimizer::property::MonotonicityMap;
+use crate::stream_fragmenter::BuildFragmentGraphState;
+use crate::TableCatalog;
+
+/// [`StreamAsOfJoin`] implements [`super::LogicalJoin`] with hash tables.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct StreamAsOfJoin {
+    pub base: PlanBase<Stream>,
+    core: generic::Join<PlanRef>,
+
+    /// The join condition must be equivalent to `logical.on`, but separated into equal and
+    /// non-equal parts to facilitate execution later
+    eq_join_predicate: EqJoinPredicate,
+
+    /// Whether can optimize for append-only stream.
+    /// It is true if input of both side is append-only
+    is_append_only: bool,
+
+    /// inequality description
+    inequality_desc: AsOfJoinDesc,
+}
+
+impl StreamAsOfJoin {
+    pub fn new(
+        core: generic::Join<PlanRef>,
+        eq_join_predicate: EqJoinPredicate,
+        inequality_desc: AsOfJoinDesc,
+    ) -> Self {
+        assert!(core.join_type == JoinType::AsofInner || core.join_type == JoinType::AsofLeftOuter);
+
+        // Inner join won't change the append-only behavior of the stream. The rest might.
+        let append_only = match core.join_type {
+            JoinType::Inner => core.left.append_only() && core.right.append_only(),
+            _ => false,
+        };
+
+        let dist = StreamJoinCommon::derive_dist(
+            core.left.distribution(),
+            core.right.distribution(),
+            &core,
+        );
+
+        // TODO: derive watermarks
+        let watermark_columns = FixedBitSet::with_capacity(core.schema().len());
+
+        // TODO: derive from input
+        let base = PlanBase::new_stream_with_core(
+            &core,
+            dist,
+            append_only,
+            false, // TODO(rc): derive EOWC property from input
+            watermark_columns,
+            MonotonicityMap::new(), // TODO: derive monotonicity
+        );
+
+        Self {
+            base,
+            core,
+            eq_join_predicate,
+            is_append_only: append_only,
+            inequality_desc,
+        }
+    }
+
+    pub fn get_inequality_desc_from_predicate(
+        predicate: EqJoinPredicate,
+        left_input_len: usize,
+    ) -> Result<AsOfJoinDesc> {
+        let expr: ExprImpl = predicate.other_cond().clone().into();
+        if let Some((left_input_ref, expr_type, right_input_ref)) = expr.as_comparison_cond() {
+            if left_input_ref.index() < left_input_len && right_input_ref.index() >= left_input_len
+            {
+                Ok(AsOfJoinDesc {
+                    left_idx: left_input_ref.index() as u32,
+                    right_idx: (right_input_ref.index() - left_input_len) as u32,
+                    inequality_type: Self::expr_type_to_comparison_type(expr_type)?.into(),
+                })
+            } else {
```

Comment:
Does this mean the expression must be like `left_col [<|<=|>|>=] right_col` rather than `right_col [<|<=|>|>=] left_col`? Here `left/right_col` means a column from left/right side of the join.

Reply:
both should be valid expression

Reply:
Got it.